File: train_readout.m
function W_out = train_readout(spike_train, target_output, lambda)
  % spike_train : [N_res x T]
  % target_output : [N_out x T]
  % lambda : régularisation ridge (ex: 1e-3)

  X = spike_train';            % [T x N_res]
  Y = target_output';          % [T x N_out]

  % Régression ridge : W = (XᵀX + λI)⁻¹ XᵀY
  XtX = X' * X;
  N = size(XtX, 1);
  W_out = ((XtX + lambda * eye(N)) \ (X' * Y))'; % [N_out x N_res]
end

File: lif/train_readout.py
import torch

def train_readout(spikes: torch.Tensor, target: torch.Tensor, lmbd: float) -> torch.Tensor:
    X = spikes.T  # [T, N_res]
    Y = target.T  # [T, N_out]

    XtX = X.T @ X
    reg = lmbd * torch.eye(XtX.size(0))
    W_out = (XtX + reg).inverse() @ X.T @ Y
    return W_out.T  # [N_out, N_res]
